### PR TITLE
feat(typescript-estree): deprecate createDefaultProgram

### LINKED ISSUE #5857: Refactor: Deprecate createDefaultProgram/shouldCreateDefaultProgram
## Overview

Back in the day, if a user didn't provide a `parserOptions.project`, `@typescript-eslint/parser` would create one behind-the-scenes for them. This was convenient but came at severe performance cost: the parser would recompute a bunch of information _for each file added_.

We wish to remove the `createDefaultProgram` option. But, there may be some users who still rely on it. See #3627.

For now, our plan will be:

1. In the upcoming major version, rename it to `deprecated__createDefaultProgram`
2. Remove it in the _next_ major version

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
BREAKING CHANGE:
Renames a part of the public typescript-estree API

## PR Checklist

- [x] Addresses an existing open issue: fixes #5857
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Renames `createDefaultProgram` to `deprecated__createDefaultProgram`, with associated `@deprecated` TSDoc tags and warnings.

### CODE REVIEW

**Comment by bradzacher** [comment]
note: marking as a `feat` so that it shows up in the changelog.

File: packages/typescript-estree/README.md [hunk 1]
```diff
@@ -215,7 +216,7 @@ interface ParseAndGenerateServicesOptions extends ParseOptions {
    * This means that if the parser encounters a file not included in any of the provided `project`s,
    * it will not error, but will instead parse the file and its dependencies in a new program.
    */
-  createDefaultProgram?: boolean;
+  deprecated__createDefaultProgram?: boolean;
```
Comment (bradzacher) [inline packages/typescript-estree/README.md:219]: in regards to the comment, might be good to clean this up and make it more obtuse to help steer people away from it?

Like

```ts
/**
 * @deprecated - this flag will be removed in the next major.
 * Do not rely on the behaviour provided by this flag.
 */
```

Comment (bradzacher) [inline packages/typescript-estree/README.md:219]: Personally I like the all-caps to make it more obnoxious and obvious
```suggestion
  DEPRECATED__createDefaultProgram?: boolean;
```

Or could go the react route 😆 

```suggestion
  DEPRECATED_DO_NOT_USE_OR_YOU_WILL_BE_FIRED__createDefaultProgram?: boolean;
```

File: packages/typescript-estree/src/parser.ts [hunk 2]
```diff
@@ -44,7 +44,8 @@ function getProgramAndAST(
       useProvidedPrograms(parseSettings.programs, parseSettings)) ||
     (shouldProvideParserServices && createProjectProgram(parseSettings)) ||
     (shouldProvideParserServices &&
-      parseSettings.createDefaultProgram &&
+      // eslint-disable-next-line deprecation/deprecation
```
Comment (bradzacher) [inline packages/typescript-estree/src/parser.ts:47]:
```suggestion
      // eslint-disable-next-line deprecation/deprecation -- will be cleaned up with the next major
```

File: packages/typescript-estree/src/create-program/createDefaultProgram.ts [hunk 3]
```diff
@@ -66,4 +69,5 @@ function createDefaultProgram(
   return ast && { ast, program };
 }
 
+// eslint-disable-next-line deprecation/deprecation
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createDefaultProgram.ts:72]:
```suggestion
// eslint-disable-next-line deprecation/deprecation -- will be cleaned up with the next major
```

**Review by bradzacher — APPROVED** [review]
looking good to me!
deprecate_dalek

File: packages/typescript-estree/README.md [hunk 4]
```diff
@@ -208,7 +208,7 @@ interface ParseAndGenerateServicesOptions extends ParseOptions {
 
   /**
    * @deprecated - this flag will be removed in the next major.
-   * Do not rely on the behaviour provided by this flag.
+   * Do not rely on the behavior provided by this flag.
```
Comment (bradzacher) [inline packages/typescript-estree/README.md:211]: image

**Review by bradzacher — APPROVED** [review]

File: packages/typescript-estree/README.md (hunk 4 above)
Comment (JoshuaKGoldberg) [reply packages/typescript-estree/README.md:211]: 😂 cspell was yelling at me!

**Pull request merged**